orca_detector/database_parser.py

```diff
@@ -1,95 +1,257 @@
+    """
+        Load the lists of files and labels for train, val, or test set.  One-hot
+        encoding of the labels is performed at this time based on the desired
+        classes specified in orca_params.CLASSES (all others are mapped to 'Other').
+        The pickled LabelEncoder() is also saved so that an inverse transform can
+        later be performed with it if desired.
+    """
+    
+    if dataset_type not in ('train', 'validate', 'test'):
+        raise ValueError('ERROR: only train, validate, and test types are supported.')
+        
+    if dataset_type == 'train':
+        indices_file = os.path.join(data_path, 'train_map.p')
+    elif dataset_type == 'validate':
+        indices_file = os.path.join(data_path, 'validate_map.p')
+    elif dataset_type == 'test':
+        indices_file = os.path.join(data_path, 'test_map.p')
+    
+    if os.path.exists(indices_file):
+        with open(indices_file, 'rb') as f:
+            flattened = pickle.load(f)
+        print('\nLoad {} dataset from {}'.format(dataset_type, indices_file))
     else:
+        raise Exception('ERROR: run database_parser.py to generate datafiles first.')
+        
+    labels, files = zip(*flattened)
 
-        all_samples = defaultdict(list)
-        for (dirpath, dirnames, filenames) in os.walk(folder):
-            filenames = [filename for filename in filenames if os.path.splitext(filename)[1].lower() == '.wav']
-            
-            if len(filenames) == 0:
-                continue
-            path, folder = os.path.split(dirpath)
-            path, label = os.path.split(path)
-            all_samples[label].extend(
-                [os.path.join(dirpath, file) for file in filenames])
-
-        train = defaultdict(list)
-        validate = defaultdict(list)
-
-        for label, files in all_samples.items():
-            if len(files) < 10:
-                continue
-                random.shuffle(files)
-            train_files = int((len(files)+1)*split_percentage)
-            train[label] = files[:train_files]
-            validate[label] = files[train_files:]
-
-        train_flattened = [[label, file]
-                           for label in train.keys() for file in train[label]]
-        validate_flattened = [[label, file]
-                              for label in train.keys() for file in validate[label]]
-
-        validate_flattened = _quantize_samples(validate_flattened)
-        train_flattened = _quantize_samples(train_flattened)
-
-        with open(train_indices_file, 'wb') as fp:
-            pickle.dump(train_flattened, fp)
-
-        with open(validate_indices_file, 'wb') as fp:
-            pickle.dump(validate_flattened, fp)
-
-    train_labels, train_files = zip(*train_flattened)
-    validate_labels, validate_files = zip(*validate_flattened)
-
-    return train_files, _onehot(train_labels), validate_files, _onehot(validate_labels)
+    return files, _onehot(labels)
+
+    
+if __name__ == '__main__':
```

Comment:
One thing to consider for this particular file.  The more flexibility we allow (or require) based on CLI args, the more chance we lose track of what conditions were used to generate a particular data file on disk.  If the logic is all documented in the checked-in code, then we know what the files contain.  This particular file shouldn't need to be run that often -- parse the directory structure, quantize the wave files, and save all that metadata.

That's why I like the idea of *not* collapsing the majority of classes to "Other" when we write this file to disk -- because we could make that decision later.  However, since it seems that doing the computations on the fly in the generator causes too much of a performance hit, maybe we consider having a two stage setup process.  The current logic -- writing out a file with metadata wouldn't need to change much.  But then a new second stage could actually write the log mel spectrogram numpy arrays to disk as hdf5 (which should allow for high-performance reloading).

Resolving for now since I think this work should be done in a separate pull request that addresses improved performance of the generators.

Reply:
Agreed, a two-steps data ETL is more reasonable here. Have in mind that we might figure out a way to optimize the preprocessing and do it in memory and avoid the disk IO.

I'll have a look in the bottleneck to see what's actually causing it, it might be something else. In any way, I agree that we're talking about two different things here, data preprocessing and data loading.

Reply:
Not sure what's the connection of this conversation with my original comment though :-)

Reply:
it was a convenient place to go off on a tangent.  Sounds like we're aligned though.